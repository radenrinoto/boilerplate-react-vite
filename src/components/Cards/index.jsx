import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import { Button } from '@mui/material';

import classes from './style.module.scss';

const Cards = ({ course, id, onClick }) => {
  const navigate = useNavigate();

  return (
    <div data-testid="cards-container" className={!course.is_close ? classes.card : `${classes.card} ${classes.disabled}`}>
      <img src={course.image} alt={course?.title} />
      <div data-testid="card-description" className={classes.cardDesc}>
        <div className={classes.desc}>
          <div className={classes.title}>{course?.title}</div>
          <div className={classes.description}>{course?.description}</div>
          <div>{course?.courseInstructor?.fullName}</div>
          <div>Rp.{course?.price?.toLocaleString()}</div>
          <div className={classes.participate}>Max Participant {course?.max_participants}</div>
        </div>
        {course?.is_close ? null : (
          <div data-testid="navigate-button" className={classes.detail} onClick={() => navigate(`/courses/${id}`)}>
            Detail Course
          </div>
        )}
        <button data-testid="button-submit" onClick={onClick}>Submit</button>
        <Button data-testid="button-mui" onClick={onClick}>BUTTON</Button>
      </div>
    </div>
  );
};

Cards.propTypes = {
  course: PropTypes.object,
  id: PropTypes.number,
  onClick: PropTypes.func
};

export default Cards;
