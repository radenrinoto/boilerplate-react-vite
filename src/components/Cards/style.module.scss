@import '@styles/base';

.card {
  width: 100%;
  background-color: white;
  border-radius: 10px;
  position: relative;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
  @include apply-at-md-size {
    @media (min-width: $screen-md-limit) {
      width: 300px;
    }
  }
  .date {
    position: absolute;
    top: 10px;
    right: 10px;
    background-color: white;
    padding: 10px;
    font-size: 12px;
    border-radius: 10px;
    font-weight: bold;
    box-shadow: 0px 0px 10px rgba($color: #000000, $alpha: 0.2);
  }
  &.disabled {
    filter: grayscale(100%);
    color: gray;
    cursor: not-allowed;
  }
}

img {
  background-position: center;
  object-position: center;
  object-fit: cover;
  width: 100%;
  height: 200px;
}

.cardDesc {
  padding: 20px;
  .title {
    color: rgb(14, 14, 66);
    font-weight: bold;
  }
  .description {
    font-size: 14px;
    margin-bottom: 10px;
  }
  .participate {
    font-size: 12px;
    font-weight: bold;
    color: gray;
  }
  .detail {
    text-align: right;
    font-size: 16px;
    font-weight: 500;
    cursor: pointer;
    &.disabled {
      cursor: not-allowed;
    }
  }
  .desc {
    margin-bottom: 20px;
  }
}
