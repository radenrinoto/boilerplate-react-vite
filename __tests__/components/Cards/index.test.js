import { render, screen, fireEvent } from '@utils/testHelper';
import Cards from '@components/Cards';

let wrapper;

let isOpen = false;

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate
}))

const mockProps = {
  title: 'Course',
  description: 'Dummy description',
  courseInstructor: {
    fullName: 'John Doe'
  },
  max_participants: 2,
  is_close: false
}

const mockId = 2;

const mockOnClick = () => {
  isOpen = true
}

beforeEach(() => {
  wrapper = render(<Cards course={mockProps} id={mockId} onClick={mockOnClick} />);
});

describe('Cards Test', () => {
  test('should render correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('cards-container')).toBeInTheDocument();
    expect(getByTestId('card-description')).toBeInTheDocument();
  })
  test('Should render text correctly', () => {
    expect(screen.getByText(`Max Participant ${mockProps.max_participants}`)).toBeInTheDocument();
  })
  test('Should call onClick when button clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-submit');
    fireEvent.click(button);
    expect(isOpen).toBe(true);
  })
  test('Should call onClick when button clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-mui');
    fireEvent.click(button);
    expect(isOpen).toBe(true);
  })
  test('Shoul call navigate', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId("navigate-button");
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/courses/${mockId}`);
  })
  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })
})