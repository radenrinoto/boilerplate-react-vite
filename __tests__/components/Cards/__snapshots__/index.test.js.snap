// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards Test Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
        data-testid="cards-container"
      >
        <img
          alt="Course"
        />
        <div
          class="cardDesc"
          data-testid="card-description"
        >
          <div
            class="desc"
          >
            <div
              class="title"
            >
              Course
            </div>
            <div
              class="description"
            >
              Dummy description
            </div>
            <div>
              John Doe
            </div>
            <div>
              Rp.
            </div>
            <div
              class="participate"
            >
              Max Participant 
              2
            </div>
          </div>
          <div
            class="detail"
            data-testid="navigate-button"
          >
            Detail Course
          </div>
          <button
            data-testid="button-submit"
          >
            Submit
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="button-mui"
            tabindex="0"
            type="button"
          >
            BUTTON
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
      data-testid="cards-container"
    >
      <img
        alt="Course"
      />
      <div
        class="cardDesc"
        data-testid="card-description"
      >
        <div
          class="desc"
        >
          <div
            class="title"
          >
            Course
          </div>
          <div
            class="description"
          >
            Dummy description
          </div>
          <div>
            John Doe
          </div>
          <div>
            Rp.
          </div>
          <div
            class="participate"
          >
            Max Participant 
            2
          </div>
        </div>
        <div
          class="detail"
          data-testid="navigate-button"
        >
          Detail Course
        </div>
        <button
          data-testid="button-submit"
        >
          Submit
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          data-testid="button-mui"
          tabindex="0"
          type="button"
        >
          BUTTON
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
